### ISSUE TITLE
[Bug]: `Component` causing incorrect re-renders

### ISSUE THREAD

**Issue by brophdawg11** [remix-run/react-router#10283]
### What version of React Router are you using?

6.9.0

### Steps to Reproduce

[Original Discord thread](https://discord.com/channels/770287896669978684/1090643300446388385/1090643300446388385)

```jsx
let router = createBrowserRouter([
  {
    path: "/",
    Component: Layout,
    children: [
      {
        index: true,
        Component: Index,
      },
      {
        path: "a",
        Component: A,
      },
    ],
  },
]);

export default function App() {
  return <RouterProvider router={router} />;
}

function Layout() {
  console.log("rendering Layout");
  return (
    <>
      <Link to="/">Home</Link>&nbsp;&nbsp;
      <Link to="a">A</Link>&nbsp;&nbsp;
      <Outlet />
    </>
  );
}

function Index() {
  console.log("rendering Index");
  return <h1>Index</h1>;
}

function A() {
  console.log("rendering A");
  return <h1>A</h1>;
}
```

### Expected Behavior

`Component` shouldn't cause additional re-renders compared to `element`

### Actual Behavior

`Component` causes additional re-renders because `React.createElement` is re-called each time

**Comment by brophdawg11** [comment]
Current workaround is to use `Component: React.memo(Layout)`

**Comment by MichaelDeBoey** [comment]
Fixed by #10287

**Comment by brophdawg11** [comment]
Reopening until this is released to npm
